src/Components/Canvas/Canvas.tsx

```diff
@@ -1,98 +1,176 @@
+            setupGif(setRecordCanvasInParent);
+            setRecordCanvasInParent(true);
+        })
+    }
+
+    function updateGif() {
+        if (save && renderer) {
+            p.save("drawgazer-screenshot");
+            save = false;
+            setSaveCanvasInParent();
+        }
+
+        if (record && renderer) {
+            if (frameCounter < fps * maxSecondsRecorded) {
+                if (frameCounter % sampleRate == 0) {
+                    gif.addFrame(p.canvas, { delay: 1, copy: true });
+                }
+                frameCounter++;
+            }
+            else if (frameCounter == fps * maxSecondsRecorded) {
+                record = false;
+                gif.render();
+                frameCounter = 0;
+                setRecordCanvasInParent(false);
+            }
+        }
+    }
+
     p.setup = function () {
         renderer = p.createCanvas(sketchData.canvasWidth, sketchData.canvasHeight);
         renderer.parent("canvas");
         sketchData.points = [];
         Animation.propsHandler(sketchData, p);
 
+        setupGif(setRecordCanvasInParent);
         settingState = 0;
     }
 
     p.windowResized = function () {
-        sketchData.canvasHeight = window.innerHeight * 0.75 -  sketchData.bufferHeight;
+        sketchData.canvasHeight = window.innerHeight * 0.75 - sketchData.bufferHeight;
         sketchData.canvasWidth = window.innerWidth * 0.85 - sketchData.bufferWidth;
         p.resizeCanvas(sketchData.canvasWidth, sketchData.canvasHeight);
     }
 
     p.myCustomRedrawAccordingToNewPropsHandler = function (props) {
         sketchData.selectedFigure = props.canvasSettings.selectedFigure;
         sketchData.selectedAnimation = props.canvasSettings.selectedAnimation;
-        
-        if (props.canvasSettings.colorSettings && 
+
+        if (props.canvasSettings.colorSettings &&
             props.canvasSettings.colorSettings != sketchData.colorSettings) {
-                sketchData.colorSettings = props.canvasSettings.colorSettings;
-                Animation.propsHandler(sketchData, p);
+            sketchData.colorSettings = props.canvasSettings.colorSettings;
+            Animation.propsHandler(sketchData, p);
         }
 
         reset = props.canvasSettings.reset;
+        save = props.canvasSettings.save;
         setClearCanvasInParent = props.canvasSettings.resetInParent;
+        setSaveCanvasInParent = props.canvasSettings.saveInParent;
+        setRecordCanvasInParent = props.canvasSettings.recordInParent;
         settingState = props.canvasSettings.settingState;
-        
+
+        if (record != props.canvasSettings.record) {
+
+            // this case means that the recording was turned off
+            if (record == true) {
+                record = false;
+                if (gif) gif.render();
+                frameCounter = 0;
+            }
+
+            record = props.canvasSettings.record;
+            setupGif(setRecordCanvasInParent);
+        }
+
```

Comment:
I don't think this is working as expected. If you stop the recording early it continues to add frames while rendering the gif.

Reply:
can you try again, this is definitely working correctly for me?

Reply:
Add a console.log() on line 115. It still prints while the gif is rendering.

Reply:
this has been fixed. However, we really shouldn't be using javascript numbers to store state any more please.